From 34d2bb917cc60e3b2e4fe5777429cf5ea164b24e Mon Sep 17 00:00:00 2001
From: Bernhard Nebel <nebel@hinterm-ziel.de>
Date: Fri, 6 Feb 2026 15:55:42 +0100
Subject: [PATCH] Mask out non-zero bits in flas addresses retrieved from the
 stack

---
 gdb/avr-tdep.c | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/gdb/avr-tdep.c b/gdb/avr-tdep.c
index fe0283f1b48..d6466baa9e2 100644
--- a/gdb/avr-tdep.c
+++ b/gdb/avr-tdep.c
@@ -38,6 +38,7 @@
 #include "objfiles.h"
 #include <algorithm>
 #include "gdbarch.h"
+#include "memattr.h"
 
 /* AVR Background:
 
@@ -240,7 +241,8 @@ avr_register_type (struct gdbarch *gdbarch, int reg_nr)
   return builtin_type (gdbarch)->builtin_uint8;
 }
 
-/* Instruction address checks and conversions.  */
+
+/* Instruction address checks and conversions.	*/
 
 static CORE_ADDR
 avr_make_iaddr (CORE_ADDR x)
@@ -1057,6 +1059,25 @@ avr_frame_unwind_cache (const frame_info_ptr &this_frame,
   return info;
 }
 
+/* Mask out unused bits in flash memory addresses in order to allow for correct
+   backtraces. This needs to be done for ATmega16 and ATmega64. For these chips, 
+   this is documented in the data sheets, but apparently happens also for other chips.
+   The code below works only when the memory map is supplied by the remote server.
+   In other words, for older servers or GDB compiled without expat, the problem
+   is unsolved, and backtraces will fail. */
+
+static CORE_ADDR
+avr_pc_mask()
+{
+  static CORE_ADDR pcmask = 0;
+  if (pcmask == 0) {
+    struct mem_region *m = lookup_mem_region(0); /* determine flash space */
+    pcmask = m->hi-1;
+  }
+  return pcmask;
+}
+
+
 static CORE_ADDR
 avr_unwind_pc (struct gdbarch *gdbarch, const frame_info_ptr &next_frame)
 {
@@ -1064,7 +1085,7 @@ avr_unwind_pc (struct gdbarch *gdbarch, const frame_info_ptr &next_frame)
 
   pc = frame_unwind_register_unsigned (next_frame, AVR_PC_REGNUM);
 
-  return avr_make_iaddr (pc);
+  return avr_make_iaddr (pc&avr_pc_mask());
 }
 
 static CORE_ADDR
-- 
2.52.0

